### PR TITLE
Add unchecked/SIMD bitshift checks and disable CBMC flag

### LINKED ISSUE #1963: `simd_shl` and `simd_shr` don't check overflow cases
Both `simd_shl` and `simd_shr` are missing an overflow check that is done for the regular bit-shift operators (i.e., `<<` and `>>`). Consider this example:

```rust
#![feature(repr_simd, platform_intrinsics)]

#[repr(simd)]
#[allow(non_camel_case_types)]
#[derive(Clone, Copy, PartialEq, Eq)]
pub struct i32x2(i32, i32);

extern "platform-intrinsic" {
    fn simd_shl<T>(x: T, y: T) -> T;
    fn simd_shr<T>(x: T, y: T) -> T;
}

#[kani::proof]
fn test_normal() {
    let value: i32 = kani::any();
    let shift: i32 = kani::any();
    kani::assume(value == 1);
    kani::assume(shift == 32);
    let result: i32 = value << shift;
}

#[kani::proof]
fn test_simd() {
    let value = kani::any();
    kani::assume(value == 1);
    let values = i32x2(value, value);
    let shift = kani::any();
    kani::assume(shift == 32);
    let shifts = i32x2(shift, shift);
    let result = unsafe { simd_shl(values, shifts) };
}
```

This is the verification output for `test_normal`:

```
RESULTS:
Check 1: test_normal.assertion.1
         - Status: FAILURE
         - Description: "attempt to shift left with overflow"
         - Location: tests/kani/Intrinsics/SIMD/Operators/bitshift.rs:29:23 in function test_normal

Check 2: test_normal.undefined-shift.1
         - Status: SUCCESS
         - Description: "shift distance is negative"
         - Location: tests/kani/Intrinsics/SIMD/Operators/bitshift.rs:29:23 in function test_normal

Check 3: test_normal.undefined-shift.2
         - Status: FAILURE
         - Description: "shift distance too large"
         - Location: tests/kani/Intrinsics/SIMD/Operators/bitshift.rs:29:23 in function test_normal

Check 4: test_normal.undefined-shift.3
         - Status: SUCCESS
         - Description: "shift operand is negative"
         - Location: tests/kani/Intrinsics/SIMD/Operators/bitshift.rs:29:23 in function test_normal
```

This is the verification output for `test_simd`:

```
RESULTS:
Check 1: undefined-shift.1
         - Status: SUCCESS
         - Description: "shift distance is negative"

Check 2: undefined-shift.2
         - Status: FAILURE
         - Description: "shift distance too large"

Check 3: undefined-shift.3
         - Status: SUCCESS
         - Description: "shift operand is negative"

Check 4: undefined-shift.4
         - Status: SUCCESS
         - Description: "shift distance is negative"

Check 5: undefined-shift.5
         - Status: FAILURE
         - Description: "shift distance too large"

Check 6: undefined-shift.6
         - Status: SUCCESS
         - Description: "shift operand is negative"
```

In https://github.com/diffblue/cbmc/issues/7362 I asked the CBMC about this missing check, but it turns out these checks are being generated by MIR! It's not clear to me if they're necessary, as some overflow cases overlap with the "shift is too large" cases, for example. It's also unclear how to include them for SIMD intrinsics.

### LINKED ISSUE #2374: Unexpected failed check for negative left shift operand
Apologies in advance if this is intended behavior! I couldn't find any previous discussion of this.

I was playing around with Kani and the chrono crate when I encountered "Failed Checks: shift operand is negative". It seems chono's internals use a packed representation for dates that involves left shifting i32 years.

I tried this code in a blank project:
```rust
fn main() {}

#[cfg(kani)]
#[kani::proof]
fn proof() {
    let n: i32 = -16;
    let s: u8 = kani::any();
    kani::assume(s == 1);
    let _ = n << s;
}
```

using the following command line invocation: `cargo kani`

with Kani version: 0.25.0

I expected to see this happen: My understanding is that left shifting a negative integer is well-defined safe behavior in Rust. I wouldn't expect there to be any failed checks (at least not without an overflow).

Instead, this happened: "Failed Checks: shift operand is negative"

### PR DESCRIPTION
**Pull request by reisnera**
### Description of changes:
The CBMB flag `--undefined-shift-check` is too strict. In particular, it flags shifting a negative value as being incorrect, which is not the case for Rust where shifts are valid unless the shift distance is negative or larger than the bit width of the value being shifted.

This PR disables the CBMC flag and adds checks to Kani both for the unchecked shift BinOps as well as SIMD shift intrinsics.

### Resolved issues:
Resolves #2374 - shifting a negative value now works.
Resolves #1963 - SIMD shifts now have the same checks as non-SIMD-shifts, namely that the shift distance is not negative or larger than the bit width of the value being shifted.

### Related RFC:
n/a

### Call-outs:
The changes in intrinsic.rs could probably be more DRY but I'm not sure it's worth it.

### Testing:
* How is this change tested?
I believe existing tests cover this PR.

* Is this a refactor change?
No

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by celinval — COMMENTED** [review]
It looks pretty good. Can you please add a test to ensure that CBMC does the expected thing for shift with negative values?

Something like:
```rust
#[kani::proof]
#[kani::unwind(32)]
fn check_shl() {
    let val : i32 = kani::any();
    let dist = kani::any_where(|d: &u8| *d < 32); 
    assert_eq!(val << dist, val.wrapping_mul(2_i32.wrapping_pow(dist.into())));
}
```
you might also want to add a modified version of the test you created in the issue description.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs [hunk 1]
```diff
@@ -455,6 +466,40 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        let value_width_expr = Expr::int_constant(value_width, distance_expr.typ().clone());
+
+        let excessive_distance_check = self.codegen_assert_assume(
+            distance_expr.clone().lt(value_width_expr),
+            PropertyClass::ArithmeticOverflow,
+            "attempt to shift by excessive shift distance",
+            loc,
+        );
+
+        // The first argument to the shift intrinsics determines the types of both arguments
```
[the remark this answers is not in the document]
Comment (reisnera) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:489]: Gotcha. I added it because I thought it might be confusing that the signedness of `value_expr` is checked but then `distance_expr` is actually used. Now that I think about it again, should it just be checking if `distance_expr` is signed instead?

[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:489]: Yeah, I think that would make more sense. Thanks!

**Comment by reisnera** [comment]
Will work on adding the tests when I can! Also, my clock somehow got out of sync so I apologize about any incorrect date/timestamps on these commits.

**Review by celinval — APPROVED** [review]
Thanks

**Review by adpaco-aws — APPROVED** [review]
Thank you, @reisnera !

> my clock somehow got out of sync so I apologize about any incorrect date/timestamps on these commits.

No worries, we use the "squash and merge" option for PRs.
